sensor/kubernetes/listener/resources/store_provider.go

```diff
@@ -56,3 +56,8 @@ func (p *InMemoryStoreProvider) RBAC() store.RBACStore {
 func (p *InMemoryStoreProvider) EndpointManager() store.EndpointManager {
 	return p.endpointManager
 }
+
+// Nodes returns the Nodes public interface
+func (p *InMemoryStoreProvider) Nodes() *nodeStoreImpl {
```

Comment:
I know that this goes against the go convention of "Accept interfaces, return structs", but I would either change this to `store.NodeStore` or I would change the implementations above. Otherwise it looks like they are different when in reality they're meant to do the same.

Reply:
+1 to this

Reply:
I see your point. Let me share mine.

There are ~three~ four options here:

1. `func (p *InMemoryStoreProvider) Nodes() store.NodeStore`

This is a problem cause we rely on the `nodeStore` interface in some places.

```
sensor/kubernetes/listener/resources/dispatcher.go:91:65: cannot use storeProvider.Nodes() (value of type store.NodeStore) as nodeStore value in argument to newNodeDispatcher: store.NodeStore does not implement nodeStore (missing method addOrUpdateNode) (typecheck)
```

2. `func (p *InMemoryStoreProvider) Nodes() nodeStore`

This is fine for now, but returning an unexported interface may limit the usability of this method in the future.

3.  `func (p *InMemoryStoreProvider) Nodes() *nodeStoreImpl`

The current state. It allows to add checks for the `impl` to make sure that it implements both interfaces:

```go
var _ nodeStore = (*nodeStoreImpl)(nil)
var _ store.NodeStore = (*nodeStoreImpl)(nil)
```

4. `func (p *InMemoryStoreProvider) Nodes() BigNodeStore`

Create new interface `BigNodeStore` (name subject to change) that would do:

```go
type BigNodeStore interface {
    nodeStore
    store.NodeStore
}
```

Based on that, I think that option 3 looks the best. What do you think?

Reply:
I think there is also a 5th option where we make `store.NodeStore` a copy of `nodeStore` but the methods from `nodeStore` (`addOrUpdateNode` etc.) will be unexported in the new bigger `store.NodeStore`.

Reply:
Following the other alternative (returning impls everywhere) would look like this:

```go
type InMemoryStoreProvider struct {
	deploymentStore *DeploymentStore  // Interface named the same as implementation - add Impl to type name?
	podStore        *PodStore
	serviceStore    *serviceStore
	rbacStore       rbac.storeImpl       // Unexported - would need to export `storeImpl`
	endpointManager *endpointManagerImpl 
	nodeStore       *nodeStoreImpl
	entityStore     *clusterentities.Store  // this is actually Impl - add Impl to type name?
}
```

As it contains a lot of renaming I would like to do this in a followup PR. Would that be okay?

Reply:
I have a simple suggestion related to (1). The only place where it seems like you should be needing the private struct instead of the public interface is when instantiating the handlers. In `sensor/kubernetes/listener/resources/dispatcher.go` you could simply do:
```go
nodeDispatcher:            newNodeDispatcher(deploymentStore, storeProvider.nodeStore, endpointManager),
```
And the dispatcher will be able to call the private store functions. 

If I'm not mistaken, when you make this change, you should be able to switch the `.Nodes()` function to the public interface without any compiler errors.

This is what we do for RBACs and Services stores. Maybe it's not the ideal way of doing things and maybe you're on to something better. But I'd opt for the easiest thing here which is just passing the internal struct to the handler. We can have a sync discussion after this PR on how we could improve this if you'd like. 

Could you give it a try?

Reply:
Add me to that sync discussion if possible 🙂 I have similar feelings as Fred but I feel they might not align with "the Go way".

Reply:
Looks like it worked great! I will be merging this in a moment